.github/workflows/main.yml

```diff
@@ -4,32 +4,34 @@ on: [push]
 
 jobs:
   build:
-
     runs-on: ubuntu-latest
 
     strategy:
       matrix:
-        node-version: ["14.x", "16.x"]
-        gatsby-version: ["^3.0.0", "^4.0.0"]
-        # exclude:
-        # - node-version: "12.x"
-        #   gatsby-version: "^4.0.0"
+        node-version: ['14.x', '16.x', '18.x']
+        gatsby-version: ['^3.0.0', '^4.0.0', '^5.0.0']
+        exclude:
+          - node-version: '14.x'
+            gatsby-version: '^5.0.0'
+          - node-version: '16.x'
+            gatsby-version: '^5.0.0'
 
     steps:
-    - uses: actions/checkout@v2
-    - name: Use Node.js ${{ matrix.node-version }}
-      uses: actions/setup-node@v1
-      with:
-        node-version: ${{ matrix.node-version }}
-    - name: Install dependencies (Gatsby ${{ matrix.gatsby-version }})
-      run: |
-        node -e "const fs = require('fs'); packageJson = JSON.parse(fs.readFileSync('package.json')); packageJson.devDependencies.gatsby = process.env.GATSBY_VERSION; fs.writeFileSync('package.json', JSON.stringify(packageJson, null, 2));"
-        yarn install --no-lockfile
-      env:
-        GATSBY_VERSION: ${{ matrix.gatsby-version }}
-    - name: Run tests
-      run: |
-        cat node_modules/gatsby/package.json | grep "\"version\""
-        yarn test
-      env:
-        CI: true
+      - uses: actions/checkout@v2
+      - name: Use Node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@v1
+        with:
+          node-version: ${{ matrix.node-version }}
+      - name: Install dependencies (Gatsby ${{ matrix.gatsby-version }})
+        run: |
+          node -e "const fs = require('fs'); packageJson = JSON.parse(fs.readFileSync('package.json')); packageJson.devDependencies.gatsby = process.env.GATSBY_VERSION; packageJson.engines.node = '>=' + process.env.NODE_VERSION; fs.writeFileSync('package.json', JSON.stringify(packageJson, null, 2));"
```

Comment:
I've added this `packageJson.engines.node = '>=' + process.env.NODE_VERSION;`, otherwise the installtion failed:  

```
The engine "node" is incompatible with this module. Expected version ">=18.0.0".
```

The [build is still failing](https://github.com/browniebroke/gatsby-source-datocms/actions) though. I think we need to tweak the versions of all gatsby packages and that might be a bit cumbersome to do. Perhaps it would be easier to drop support for Gatsby <5?

Reply:
I've extracted the logic to patch versions into a script because the script in the YAML:

- was getting hard to read
- was not easy to test locally

This way I think I managed to keep Gatsby v3. The installation works but the tests fail, but since it's already broken on the same on the main branch, fixing them is getting out of scope of this pull request.